rendition: Add API Documentation (#16)
ToontownRewritten/api-doc: toon-rendition.md

# Toontown Rewritten Rendition API

## Overview

This API generates a screenshot of a toon based on a DNA string.

## Requests

An HTTP GET to the rendition endpoint, https://rendition.toontownrewritten.com/render/{dna}/{pose}/{width}x{height}.{extension}, will return the generated render.

## Parameters

| Name        | Details |
|-------------|---------|
| dna | The DNA string for the toon to be rendered. (Use Shift+F11 in-game to copy your toon DNA) |

> | pose   | Pose your toon makes. At the moment these are the 8 options. (**head, laffmeter, portrait, portait-sleep, portrait-delighted, portrait-surprise, portrait-thinking, portrait-birthday, portrait-fall, portrait-grin, cake-topper, crying, waving**) |

Reviewer: Oh and it might be worth mentioning that the DNA string can also be retrieved from the [local API](https://github.com/ToontownRewritten/api-doc/blob/master/local.md#get-toonjson).

# Toontown Rewritten Rendition API

## Overview

> This API provides a screenshot of your a toon generated from a DNA string.

Reviewer 2: ```suggestion
This API generates a screenshot of a toon based on a DNA string.
```

## Requests

An HTTP GET to the rendition endpoint, https://rendition.toontownrewritten.com/render/{dna}/{render-type}/{width}x{height}.{extension}, will return the generated render.

## Parameters

| Name        | Details |
|-------------|---------|

> | dna | The DNA string of your toon generated. |

Reviewer 2: This phrasing seems strange to me.  Would it be correct (and clearer?) to say something like "The DNA string for the toon to be rendered"?

Author: Fair point, since the OAUTH for Toon information is still confidential for partners atm.

# Toontown Rewritten Rendition API

## Overview

This API generates a screenshot of a toon based on a DNA string.

## Requests

An HTTP GET to the rendition endpoint, https://rendition.toontownrewritten.com/render/{dna}/{pose}/{width}x{height}.{extension}, will return the generated render.

## Parameters

| Name        | Details |
|-------------|---------|
| dna | The DNA string for the toon to be rendered. (Use Shift+F11 in-game to copy your toon DNA) |
| pose   | Pose your toon makes. At the moment these are the 8 options. (**head, laffmeter, portrait, portait-sleep, portrait-delighted, portrait-surprise, portrait-thinking, portrait-birthday**) |
| width   | Width of the returned render (ex. **1024**). |
| height   | Height of the returned render (ex. **1024**). |

> | extension   | File extension can be either **png** or **webp** |

Reviewer 3: Instead of using bold for pose/width/height/extension, perhaps use codeblocks, in order to be consistent with the other files. There should also be a "valid options" column with the proper options, like the example below.

| Name        | Details | Valid Options |
|-------------|---------|---------------|
| pose | The pose your toon makes. | `head`, `laffmeter`, `portrait`, `portrait-sleep`, `portrait-delighted`, `portrait-surprise`, `portrait-thinking`, `portrait-brithday`

## Response

A toon from the provided DNA string will be generated.

> Example: https://rendition.toontownrewritten.com/render/7424010101021d021d8b1b160016160104001c0000040000000000000000000000/portrait-birthday/1024x1024.png

Reviewer 3: Perhaps also include the possible errors? So far I found

- `500 Internal Server Error` whenever the DNA string is invalid.
- `404 Unknown scene` whenever you specify an invalid pose (could also be the actual parameter name?)
- `400 Unsupported resolution` whenever you do an unsupported resolution (ex. `1x1024`).
 - There's also a `404 Not Found` whenever you type an invalid resolution (ex. `aaa`).
- `400 Unsupported file type` whenever you do an unsupported file type (ex. `.jpg`).

As well as specify that it will be a `200` for the actual response

Reviewer 2: I don't know if the example rendition is supposed to work or if it's purely an example, but it doesn't render anything.

Reviewer 3: > I don't know if the example rendition is supposed to work or if it's purely an example, but it gives an internal server error.

The example rendition works fine for me, this is the result as of 2024-05-17:
![image](https://github.com/ToontownRewritten/api-doc/assets/57161803/5b858480-2212-4aa9-9d65-a234de1636e0)

Author: The day you tested this out there was a widespread outage for the TTR rendition site API. Everything is working a-ok now
